### ISSUE TITLE
Global Serverless plugins support

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#3319]
# This is a Feature Proposal

## Description

Right now Serverless is only able to load plugins which are installed via `npm` and added to the `plugins` array of the `serverless.yml` file.

In the future Serverless should be able to detect globally available Serverless plugins and include them by default without the need to add them to the `plugins` array in `serverless.yml`.

This issue discusses different approaches how this could be implemented.

/cc @nikgraf @brianneisler @eahefnawy @mthenw 

---

## 1. Auto detect globally installed Serverless plugins

User installs the plugin via `npm install -g serverless-plugin-foo`.

Serverless looks into the global `node_modules` directory and determine Serverless plugins. Autoload those.

**Upsides** 👍
- Easy to implement?!
- Utilize npms default storage / less fragile in the future

**Downsides** 👎
- Can be super slow since we need to determine if the npm package is a Serverless plugin (and we need to do this for every globally installed package)
- Detection of Serverless plugins can be unreliable
- Possible surprising side-effects for users
- Hard to configure since ALL Serverless plugins are loaded by default

## 2. Auto load Serverless plugins which are in the homedir `.serverless` directory

User moves the plugin to `~/.serverless`.

Serverless checks the `~/.serverless` directory and loads all present plugins in there.

**Upsides** 👍 
- Easy to implement (trivial loading of plugins)
- User has way more control which plugin should be loaded by default

**Downsides** 👎 
- User needs to explicitly specify the location when he installs a plugin
- Harder to update / maintain the plugin for the user

**Comment by HyperBrain** [comment]
In my opinion auto loading of plugins is not a good idea at all. Consider you have a plugin that you do not want to or even can enable on each project (e.g. webpack) but want to install it globally on the build server to make sure that a defined version is used. Also you'd have to take extreme care to keep your node_modules dir clean locally - otherwise remnants of tests could trigger autoload of any not wanted plugin still present in the node_modules directory.

Auto loading would break the builds in this case. The (not merged) plugin autoloader in SLS 0.5 is a good example where you can see what happens.

If autoloading is integrated, I strongly propose that there is a switch where you can enable it (maybe per project).

**Comment by pmuens** [comment]
@HyperBrain thanks for the quick, valuable feedback!

> If autoloading is integrated, I strongly propose that there is a switch where you can enable it (maybe per project).

That's a really good idea! Could be a property in `serverless.yml` 👍

**Comment by HyperBrain** [comment]
@pmuens It would be quite easy to make this a very descriptive setting in serverless.yml.
We just could introduce a special plugin "name" for that, that has to be set in the plugins property. This might even be a plain string (opposed to an array entry) so that the serverless plugin loader can easily switch modes:

With autoloader enabled:
```
...
plugins: "autoloader"
...
```

With autoloader disabled:
```
...
plugins:
  - serverless-webpack
  - serverless-offline
...
```

**Comment by pmuens** [comment]
@HyperBrain neat idea! I like that.

How do we deal with the case where I want to use autoloader in combination with a list of plugins?

Or could we make this `opt out` since the user who uses globally installed plugins wants to use them and knows what he's doing.

Maybe something like

```yml
useGlobalPlugins: false # defaults to true
```

or

```yml
globalPlugins: false
```

**Comment by HyperBrain** [comment]
What if the autoloader is itself implemented as core plugin?
Then you can just specify:
plugins:
  serverless-plugin-autoloader

And you could add an arbitrary list of additional plugins after the autoloader.

**Comment by horike37** [comment]
Is this issue related to https://github.com/serverless/serverless/issues/2861?
I want the official plugin repo like https://wordpress.org/plugins/.

My idea is the following about install process.
- Run `sls plugin install <someplugin>`, so downloaded it from the official repo.
- Installed the plugin under `.serverless` or `.serverless-plugin?`.
- Autoload this.

I think the best that it can be managed by `the plugin command` and the official repo.

**Comment by pmuens** [comment]
@horike37 thanks for the reply! 👍 

> Is this issue related to #2861?

It's not directly related, but we've reopened #2861 since it's way nicer to install and discover plugins through the CLI (like you already mentioned). Thanks for the proposed workflow how this can look like!

The global plugin support should provide a way that users can auto-add often used plugins to their projects w/o having to add them manually every time.

**Comment by horike37** [comment]
Oh I see.
 I'm in favor of option 2 since  it seems to be easier for users and developers to maintain when separating them from node_module. it seems to be complicated to include in node_modules.

For example, it is better that you run `sls plugin install <plugin name> -g`, then directly installed in the homedir `.serverless` directory.

**Comment by andymac4182** [comment]
Personally I think global plugins could cause lots of weird issues that will be hard to diagnose. Things like different versions on each developers machines and projects that need certain versions. We have moved to installing serverless as a dev dependancy for the same reason. When working on multiple projects you don't always have time to keep them at the same version with breaking changes.

**Comment by arabold** [comment]
I'd like to iterate on @andymac4182's comment as I think it can get even worse than that: Serverless 1.5, 1.6, 1.7 and 1.8 have incompatibilities (breaking changes) between them that require code changes in both your application and Serverless plugins. There are actual differences in deployment when using 1.5 vs 1.6 and once upgraded older installations don't work the same way anymore.

At this point I don't think it's practical to use globally installed plugins. Not even a global Serverless installation can really be recommended until stricter versioning is introduced. This is a hell to maintain in a professional environment with multiple engineers working on the same code base. In combination with an autoloader this will unavoidably lead to disaster. In fact, I'd even go so far and update the official documentation to recommended installing Serverless and all plugins as `devDependency` in `package.json` and never globally.

**TL;DR.** Global (plugin) installation is evil and should be avoided.

![image](https://cloud.githubusercontent.com/assets/1548229/23580548/98fe7f04-00c9-11e7-9098-f60b1a30f4a0.png)

**Comment by nikgraf** [comment]
As pointed out a couple times implicit dependancies could lead to some unexpected side-effects. Nevertheless global usable plugins have some value.

The two issues I would like to see solved:
1. I want to add commands to `serverless` globally e.g custom template generation
2. I want to avoid plugins being packaged together with my code when using node e.g. the Openwhisk plugin has 40mb if installed locally in the service this would deployed with it.

Right now plugins are only used if defined inside the service. I would suggest we keep it that way if inside a service. This way we can avoid unexpected side-effects if working inside a service directory with a `serverless.yml`.

## Solution 3

This would solve only issue 2.

To make sure plugins are still controlled by semver we extend the syntax in `serverless.yml`:

```yaml
service: my-test-service

plugins:
  serverless-offline:
    version: ^2.0.0 # semver supported
  serverless-dynamodb:
    version: =1.2.2
```

In case no version is defined it accepts any version. We still accept the old list based plugin configuration, but plan to deprecate it in a couple months. If defined serverless will look first into the local node_modules and then the global location for the respective plugin.

## Solution 4

This would solve issue 1 & 2.

We use the same syntax change in `serverless.yml` as in Solution 3. If inside a directory with a `serverless.yml` we only activate the plugins defined in the `plugins` list/object. When you run `serverless` it would show that it dedected more global plugins and their commands, but they are not active.

Inside a service directory with a `serverless.yml` but without the plugins being added:
```
➜  ~ serverless

Commands
* Serverless documentation: http://docs.serverless.com
* You can run commands with "serverless" or the shortcut "sls"
* Pass "--help" after any <command> for contextual help

config credentials ............ Configures a new provider profile for the Serverless Framework
create ........................ Create new Serverless service
install ....................... Install a Serverless service from GitHub
deploy ........................ Deploy a Serverless service
deploy function ............... Deploy a single function from the service
deploy list ................... List deployed version of your Serverless Service
invoke ........................ Invoke a deployed function
invoke local .................. Invoke function locally
info .......................... Display information about the service
logs .......................... Output the logs of a deployed function
metrics ....................... Show metrics for a specific function
remove ........................ Remove Serverless service and all resources
rollback ...................... Rollback the Serverless service to a specific deployment
slstats ....................... Enable or disable stats
offline (deactive) ............ Run a localhost version
dynamodb (deactive) ........... Run dynamodb locally

Active Plugins
AwsConfigCredentials, Config, Create, Deploy, Info, Install, Invoke, Logs, Metrics, Package, Remove, Rollback, SlStats

Deactivated Plugins
serverless-offline, serverless-dynamodb
```

Not inside a service directory with a `serverless.yml` but without the plugins being added:
```
➜  ~ serverless

Commands
* Serverless documentation: http://docs.serverless.com
* You can run commands with "serverless" or the shortcut "sls"
* Pass "--help" after any <command> for contextual help

config credentials ............ Configures a new provider profile for the Serverless Framework
create ........................ Create new Serverless service
install ....................... Install a Serverless service from GitHub
deploy ........................ Deploy a Serverless service
deploy function ............... Deploy a single function from the service
deploy list ................... List deployed version of your Serverless Service
invoke ........................ Invoke a deployed function
invoke local .................. Invoke function locally
info .......................... Display information about the service
logs .......................... Output the logs of a deployed function
metrics ....................... Show metrics for a specific function
remove ........................ Remove Serverless service and all resources
rollback ...................... Rollback the Serverless service to a specific deployment
slstats ....................... Enable or disable stats
offline ....................... Run a localhost version
dynamodb ...................... Run dynamodb locally

Active Plugins
AwsConfigCredentials, Config, Create, Deploy, Info, Install, Invoke, Logs, Metrics, Package, Remove, Rollback, SlStats, serverless-offline, serverless-dynamodb
```

## Global Location

For both solutions I would rather use the global npm. Then people can choose if they want to use yarn or npm, but the other option is also fine. We could even go with both. When choosing npm plugins are only dedected if they prefix with `serverless-`.

**Comment by nikgraf** [comment]
btw in general I agree with @arabold. If serverless would just be local that would be best! Unfortunately then we don't have CLI commands available and we need a waaay smarter packaging mechanism since the whole serverless codebase would be uploaded. I tried webpack a couple times, but always ran into issues with some dependencies and then dropped it again … In summary I'm thinking about this regularly, but the solutions above are the best I could come up with so far … 

@arabold I highly recommend you to lock down the serverless framework version using the `frameworkVersion` property in `serverless.yml` e.g. what I do is `frameworkVersion: "=1.8.0" to avoid any surprises.

**Comment by andymac4182** [comment]
@nikgraf Locking down the `frameworkVersion` doesn't help if you are working on multiple projects with different versions. We currently have multiple projects and adding more regularly and while most are on a single version of serverless, some old projects that haven't been worked on for a while aren't up to date. Local serverless allows us to not worry about if the serverless command is available on the server because `yarn` will handle that when restoring packages.

Personally any commands that should be global should be split out into a `serverless-cli` package that doesn't increment often. This can have its own plugins that add global commands that aren't relevant to a individual project such as install and create. 

Defining the version of the plugin in `serverless.yml` runs into the same issue as locking down the `frameworkVersion` in that if you need a different version for another project you will have to keep changing your global. Since lots of projects don't follow SemVer strictly using `^` in a package version can end badly if you have different versions on your machine vs your CI vs another developers machine. We also have a method of defining a version for a project now in `package.json` that supports the syntax you were proposing.

While I know there have been a few packages that conflict with webpack using webpack + babel has been great to make our services smaller and easier to deploy. There is so much extra junk in the `node_modules` folder other than just serverless plugins.

**Comment by nikgraf** [comment]
@andymac4182 How did you resolve the webpack issues or do you use other packages then?

Totally agree on not using `^` as it causes more trouble then it helps. I always use `=`, but some people really really really want this (I don't know why) 🙂 

@andymac4182 With the solution proposal 3 or 4 and you using serverless and all plugins as local packages this change would not harm your existing workflow. Is this correct or do I miss something?

**Comment by gertjvr** [comment]
serverless is over reaching in this case, serverless cli to create / install new services can be globally installed :+1: serverless plugins which is a per project concern installed in a global case can cause unwanted / unexpected behavior in other unrelated projects. 👎 look at `create-react-app` they have a cli which installs the app but in installs locally `react-scripts` which is all you need. Look at other echo systems, plugins are always a local concern.

**Comment by gertjvr** [comment]
And lastly respect my development environment, don't litter my environment with unwanted and needed software installed in unrelated locations that I need to remember to clean out when I don't work on serverless project any more.

**Comment by johncmckim** [comment]
@pmuens can I ask what is the use case behind this idea? What problem is this trying to solve? I see only downsides (well documented above) and no upside to global plugins.

**Comment by pmuens** [comment]
First of all. Thanks for all the feedback so far. It's really valuable to have such a vibrant and engaged community.

> Can I ask what is the use case behind this idea? What problem is this trying to solve?

The main idea is to introduce temporary plugins which can be e.g. used as a help when breaking changes are introduced or when features which are not in core should be used.

**Comment by marckaraujo** [comment]
For me as it clearly shows, this feature request isnt too important to community right now to be target to 1.9 release. I can target a lot of others features requests that is more important to be discussed and implemented at moment. Maybe a topic about what is the next feature to be implemented or a roadmap list should be created.

**Comment by pmuens** [comment]
Closing this issue since we're currently not working on a global plugin support anymore.

Thanks everyone for chiming in and providing detailed feedback!

**Comment by Jeff-Tian** [comment]
> Closing this issue since we're currently not working on a global plugin support anymore.
> 
> Thanks everyone for chiming in and providing detailed feedback!


I want to deploy my aws lambda by serverless, the repo is here: https://github.com/Jeff-Tian/serverless-space.

Basically it's a nest serverless app, and I configured my serverless.yml to use layer manager so I can put my node_modules into that layer.

I installed the production-only node_modules so the size is only 93M. But if I ran serverless deploy then the error occurs, claiming my unzipped size exceeds 250M.

Then I checked my node_moudles size again by du -h -d 0 node_modules, it shows it's 400+ M now.

So I guess the serverless deploy installed extra packages to my node_modules, and can be fixed by install serverless plugins globally. But unfortunately there is no this kind of option accoring to this thread: https://github.com/serverless/serverless/issues/3319#issuecomment-303080193.

Then how can I workaround my deployment issues?

**Comment by mnapoli** [comment]
@Jeff-Tian please open a separate issue.

**Comment by Jeff-Tian** [comment]
> @Jeff-Tian please open a separate issue.

I will skip the new issue creating as I fixed the issue by not using `serverless plugin install plugin-name`, instead, I use

```
npm i -g serverless
npm i -g serverless-plug-in-name
```

So I don't need the serverless itself to support global plugin. 💯
